File: test/BallRecognition.py
import numpy as np
import cv2
import time
import datetime


def detect(frame):
    """
    :param frame: 640*480 jpeg image
    :return: x coordinate of ball in image
    """
    hue_image = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)  # RGB色彩转化为HSV
    lower_red = np.array([0, 127, 128])
    upper_red = np.array([10, 255, 240])
    mask0 = cv2.inRange(hue_image, lower_red, upper_red)
    # upper mask (170-180)
    lower_red = np.array([150, 80, 50])
    upper_red = np.array([179, 229, 229])
    mask1 = cv2.inRange(hue_image, lower_red, upper_red)

    # join my masks

    mask = mask0 + mask1
    dilated = cv2.dilate(mask, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3)), iterations=1)
    cnts = cv2.findContours(mask.copy(), cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)[-2]

    if len(cnts) > 0:
        c = max(cnts, key=cv2.contourArea)
        ((x, y), radius) = cv2.minEnclosingCircle(c)
        circle_x = int(x)
        circle_y = int(y)
        if radius > 15:
            cv2.circle(frame, (int(x), int(y)), int(radius), (0, 255, 255), 2)
            return circle_x
        else:
            return -1
    else:
        return -1

File: test/ballrecognition.js
function ball(cv, frame) {
    let mat = cv.imread(frame);
    let gray = new cv.Mat();
    let dst = new cv.Mat();
    cv.cvtColor(mat, gray, cv.COLOR_BGR2HSV)

    let lower_red = new cv.Mat(mat.rows, mat.cols, mat.type(), [0, 0, 10, 0]);

    let upper_red = new cv.Mat(mat.rows, mat.cols, mat.type(), [170, 160, 160, 255]);

    cv.inRange(mat, lower_red, upper_red, dst);
    cv.bitwise_not(dst, dst)
    //console.log(dst)
    cv.dilate(dst, gray, cv.getStructuringElement(cv.MORPH_ELLIPSE, new cv.Size(5, 5), new cv.Point(-1, -1)), new cv.Point(-1, -1), 1, cv.BORDER_CONSTANT, cv.morphologyDefaultBorderValue())

    let contours = new cv.MatVector();
    let hierarchy = new cv.Mat();
    console.log(cv.RETR_EXTERNAL)
    console.log(gray)
    //  gray= new Uint8Array(gray)
    console.log(gray)
    cv.findContours(gray, contours, hierarchy, cv.RETR_CCOMP, cv.CHAIN_APPROX_SIMPLE)

    console.log(contours.size())
    var prev_c = 0
    var index = 0
    lower_red.delete();
    upper_red.delete();
    mat.delete();
    gray.delete();
    dst.delete();
    contours.delete();
    hierarchy.delete();
    if (contours.size() > 0) {
        for (let i = 0; i < contours.size(); i++) {
            var cnt = contours.get(i)
            let c = cv.contourArea(cnt, false)
            if (i == 0) {
                prev_c = c
                index = i
            }
            if (c > prev_c) {
                prev_c = c
                index = i
            }
        }

        console.log(prev_c)
        console.log(index)
        let circle = cv.minEnclosingCircle(contours.get(index))
        console.log(circle.center.x)
        // let circleColor = new cv.Scalar(255, 255, 0)
        // cv.circle(gray, circle.center, circle.radius, circleColor);
        // cv.imshow('canvasOutput', gray)

        return circle.center.x
    } else {
        return -1
    }

}
